### PR TITLE
Issue 279 - Add good compute-to-data example README

### LINKED ISSUE #279: Add good compute-to-data example README
Pre-requisite: #101

Motivation:
- Help stabilize compute-to-data
- Nice onboarding for python users to (a) use compute-to-data, in (b) a data science context

Basic flow of README
- Alice mints datatokens and publishes data asset
- transfers some to Bob
- Bob consumes in a CtD setting. The algorithm can be simple, eg `average()`

What it does _not_ need, and should not have
- Marketplace stuff
- Conversion between fiat value and OCEAN
- AI/ML tool (eg logistic regression from scikit-learn), since Raven is already working on that.

Potential sources of useful info:
- https://github.com/oceanprotocol/ocean.js-cli/blob/main/src/commands.ts
- https://github.com/oceanprotocol/ocean.py/blob/main/ocean_lib/ocean/test/test_ocean_compute.py
- #101 has a flow shared by Alex C
- simple README, marketplace README
- 20210429 Matthias list of CtD resources: https://oceanprotocol.slack.com/archives/C0D6JB45D/p1619638827175700
- Ocean docs "Writing Algorithms for Compute to Data" [link](https://docs.oceanprotocol.com/tutorials/compute-to-data-algorithms/)
- [ocean_lib/data_service_provider::start_compute_job()](https://github.com/oceanprotocol/ocean.py/blob/main/ocean_lib/data_provider/data_service_provider.py)
- [test_start_compute_job_fails_empty(), more](https://github.com/oceanprotocol/ocean.py/blob/ae45e4207f8c268c8d7a9d28704663de152d4608/ocean_lib/data_provider/test/test_data_service_provider.py)
- https://docs.oceanprotocol.com/tutorials/compute-to-data-algorithms/

### PR DESCRIPTION
**Pull request by trentmc**
Fixes #279.

### CODE REVIEW

**Comment by mariacarmina** [comment]
Got the issue related `start_compute_job`.
Screenshot from 2021-09-10 17-02-58

**Comment by mariacarmina** [comment]
**UPDATE**: To use the algorithm, I need to create a dockerized algorithm and to push it to the DockerHub repo that contains the published algorithms.

**Comment by trentmc** [comment]
I was going through testing it, and ran into this issue in step 2:

```python
from ocean_lib.web3_internal.currency import to_wei
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'ocean_lib.web3_internal.currency'
```

Perhaps David's recent changes around Decimal / wei were not yet released? If yes, the fix is easy: [do the release](https://github.com/oceanprotocol/ocean.py/blob/main/READMEs/release-process.md). Then let me know, I will test again. @calina-c 


Thanks!

**Comment by mariacarmina** [comment]
> I was going through testing it, and ran into this issue in step 2:
> 
> ```python
> from ocean_lib.web3_internal.currency import to_wei
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> ModuleNotFoundError: No module named 'ocean_lib.web3_internal.currency'
> ```
> 
> Perhaps David's recent changes around Decimal / wei were not yet released? If yes, the fix is easy: [do the release](https://github.com/oceanprotocol/ocean.py/blob/main/READMEs/release-process.md). Then let me know, I will test again. @calina-c
> 
> Note that my testing was on purpose _not_ in the ocean.py codebase, but separate. To mimic how a user would see it. I'm guessing that David's changes were merged into this branch, so if you're only testing from the branch then you wouldn't have seen the error.
> 
> Also if this is the case, then quickstart marketplaces flow would break too, since it also uses to_wei.
> 
> Thanks!

Until the release, you can try using `pip install --no-cache-dir ../ocean.py/` and make sure the main branch is updated, instead of `pip install ocean-lib`.

**Comment by trentmc** [comment]
> Until the release, you can try using ...

Thanks. However to test as well as possible, I aim to replicate the true flow as directly as possible. Releasing is not a huge task. I'll wait.

**Comment by calina-c** [comment]
Pasting here for updates too: Trent's current algorithm is a simple Python Hello World, without any particular dependencies. The c2d readme now works, and the job outputs "hello world", as expected. We do not need a customised docker image **_yet_**

As I see it we have 2 choices:
- postpone the completion of the c2d readme until your algorithm is finished and we can create the custom docker image, using that inside the readme
- release an intermediary c2d readme, as it is now on a simple Python image (I addeed a reference to creating custom images -> a link to the tutorial docs — not perfect), and update it when you update the algorithm itself

While this is not perfect, it is also true that we need to update this readme once we have the computeResult endpoint integrated, so either way we’ll have to release an intermediary version.

I’m leaning towards releasing the readme, and creating a new issue for changes related to computeResult + creating a new docker image to support your needs. We also need to add some more explicit code in the tutorials, as to how people can create their custom docker images, but that is also another issue/ticket.

I'm currently discussing this with Trent.

**Comment by trentmc** [comment]
Re: Pasting here for updates too: Trent's current algorithm is a simple Python Hello World, without any particular dependencies. The c2d readme now works, and the job outputs "hello world", as expected. We do not need a customised docker image yet

Fyi the algorithm is there, it is just in another file: https://github.com/trentmc/branin/blob/main/branin.py. See the gpr building code. I had taught this to Maria when she took over the task, sorry Calina for not onboarding you on this.

To move forward, I recommend:
- play with branin.py. 
- move appropriate code to gpr.py
- do the rest

**Comment by calina-c** [comment]
The example now works with a custom algorithm URL based on my PR in branin. Once we merge that, and the updated gpr.py replaces the "hello world" one, it seems to be working like a charm.

I added some follow-up for tutorial improvements in https://github.com/oceanprotocol/docs/issues/770

And I would like to wait to integrate the computeResult endpoint (https://github.com/oceanprotocol/ocean.py/pull/470) which only depends on a barge release (which in turn depends on updating one ocean.js query). It will take another couple of hours, but I'd rather not confuse users by using one flow now, and another one next week. The changes will be minimal once a new ocean.py (one that supports result_file) is released.

**Comment by calina-c** [comment]
I also made an image processing sample: https://github.com/calina-c/ocean-lena/blob/main/c2d-flow.md

Maybe we can offer it as an alternative to the "main" c2d flow. I know some users can be more impressed by visual stuff, so image processing can be a good avenue to do so.

**Comment by trentmc** [comment]
Branin pr is merged.

I think it is useful if you give a bonus section at the end for the image stuff.

Please lmk when you want me to review.

File: READMEs/c2d-flow.md [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+from ocean_lib.data_provider.data_service_provider import DataServiceProvider
+provider_url = DataServiceProvider.get_url(ocean.config)
+# returns "http://localhost:8030"
+
+# Calc DATA service compute descriptor
+from ocean_lib.common.agreements.service_factory import ServiceDescriptor
+DATA_compute_service_descriptor = ServiceDescriptor.compute_service_descriptor(DATA_service_attributes, provider_url)
+
+#Publish metadata and service info on-chain
+DATA_ddo = ocean.assets.create(
```
Comment (akshay-ap) [inline READMEs/c2d-flow.md:176]: FYI: I got error here when running with Barge with `Aquarius` v3.1.1, But when I downgraded `Aquarius` to v3.0.1, the same instruction worked.
Error said the Datatoken was already associated with an asset although that was not true. Unfortunately, I dont have exact logs now. I can reproduce them if needed,

**Comment by calina-c** [comment]
From the branin.py repo, todo:

Put plotting into the tail end of the readme flow itself. It will give a nice visual output to people going through the flow and make their brains happy. If we keep it contact maybe it's another 10 lines of code.

Specific steps, for tail end of readme:

- generate the input 'X'. Like how branin.py does it
- run X through the recently retrieved gpr model, to get predictions 'yhat'. The branin repo has example too
- do a surface plot of X to that, like branin repo too.

File: READMEs/c2d-flow.md (hunk 1 above)
Comment (calina-c) [reply READMEs/c2d-flow.md:186]: That was ocean.py not being in sync with the Aquarius API. Fixed now with the new release.

**Comment by calina-c** [comment]
@trentmc I think everything is handled now. We can release this example when you approve the changes (via comment, I guess. You can't approve the PR if you're the original author)

File: READMEs/c2d-flow.md [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+
+# Create Alice's wallet
+import os
+from ocean_lib.web3_internal.wallet import Wallet
+alice_wallet = Wallet(ocean.web3, os.getenv('TEST_PRIVATE_KEY1'), config.block_confirmations)
+print(f"alice_wallet.address = '{alice_wallet.address}'")
+
+# Mint OCEAN for ganache only
+from ocean_lib.ocean.mint_fake_ocean import mint_fake_OCEAN
+mint_fake_OCEAN(config)
```
Comment (trentmc) [inline READMEs/c2d-flow.md:135]: You don't need this. This is only in the market README as liquidity for the AMMs. This README doesn't have that.

File: READMEs/c2d-flow.md [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+Full details: [DATA_ddo](DATA_ddo.md)
+
+
+## 3. Alice publishes algorithm
+
+For this step, there are some prerequisites needed. If you want to replace the sample algorithm with an algorithm of your choosing, you will need to do some dependency management.
+You can use one of the standard [OCEAN algo_dockers images](https://github.com/oceanprotocol/algo_dockers) or publish a custom docker image.
+Use the image name and tag in the `container` part of the algorithm metadata.
+This docker image needs to have basic support for dependency installation e.g. in the case of Python, OS-level library installations, pip installations etc.
+Take a look at the [OCEAN tutorials](https://docs.oceanprotocol.com/tutorials/compute-to-data-algorithms/) to learn more about docker image publishing.
```
Comment (trentmc) [inline READMEs/c2d-flow.md:205]: "OCEAN tutorials" => "Ocean tutorials". Only use "OCEAN" in reference to the token.

File: READMEs/c2d-flow.md [hunk 4]
```diff
@@ -0,0 +1,398 @@
...
+  publisher_wallet=alice_wallet,
+  service_descriptors=[ALG_access_service_descriptor],
+  data_token_address=ALG_datatoken.address)
+print(f"ALG did = '{ALG_ddo.did}'")
+```
+
+Full details: [ALG_ddo](ALG_ddo.md)
+
+Please note that this example features a simple Python algorithm. If you publish an algorithm in another language, make sure you have an appropriate container to run it, including dependencies.
+You can find more information about how to do this in the [OCEAN tutorials](https://docs.oceanprotocol.com/tutorials/compute-to-data-algorithms/).
```
Comment (trentmc) [inline READMEs/c2d-flow.md:267]: "OCEAN tutorials" => "Ocean tutorials".

File: READMEs/c2d-flow.md [hunk 5]
```diff
@@ -0,0 +1,398 @@
...
+bob_wallet = Wallet(ocean.web3, os.getenv('TEST_PRIVATE_KEY2'), config.block_confirmations)
+print(f"bob_wallet.address = '{bob_wallet.address}'")
+
+# Verify that Bob has ganache ETH
+assert ocean.web3.eth.get_balance(bob_wallet.address) > 0, "need ganache ETH"
+
+from ocean_lib.models.btoken import BToken #BToken is ERC20
+OCEAN_token = BToken(ocean.web3, ocean.OCEAN_address)
+assert OCEAN_token.balanceOf(alice_wallet.address) > 0, "need OCEAN"
+assert OCEAN_token.balanceOf(bob_wallet.address) > 0, "need ganache OCEAN"
```
Comment (trentmc) [inline READMEs/c2d-flow.md:291]: You don't need these lines, because you don't need OCEAN in this tutorial

File: READMEs/c2d-flow.md [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+
+1. Setup
+2. Alice publishes data asset
+3. Alice publishes algorithm
+4. Alice allows the algorithm for C2D for that data asset
+5. Bob acquires datatokens for data and algorithm
+6. Bob starts a compute job
+7. Bob monitors logs / algorithm output
+
+This c2d flow example features a simple algorithm from the field of ML. OCEAN c2d is not limited to ML datasets and algorithms, but it is one of the most common use cases. For a simple image processing example, you can check out [ocean-lena](https://github.com/calina-c/ocean-lena/blob/main/c2d-flow.md).
```
Comment (trentmc) [inline READMEs/c2d-flow.md:20]: "OCEAN c2d" -> "Ocean c2d"

File: READMEs/c2d-flow.md [hunk 7]
```diff
@@ -0,0 +1,384 @@
...
+
+In the work console:
+```console
+#set private keys of two accounts
+export TEST_PRIVATE_KEY1=0x5d75837394b078ce97bc289fa8d75e21000573520bfa7784a9d28ccaae602bf8
+export TEST_PRIVATE_KEY2=0xef4b441145c1d0f3b4bc6d61d29f5c6e502359481152f869247c7a4244d45209
+
+#needed to mint fake OCEAN for testing with ganache
+export FACTORY_DEPLOYER_PRIVATE_KEY=0xc594c6e5def4bab63ac29eed19a134c130388f74f019bc74b8f4389df2837a58
+
```
Comment (trentmc) [inline READMEs/c2d-flow.md:103]: can remove this

File: READMEs/c2d-flow.md [hunk 8]
```diff
@@ -0,0 +1,384 @@
...
+print(f"DATA did = '{DATA_ddo.did}'")
+```
+
+Full details: [DATA_ddo](DATA_ddo.md)
+
+
+## 3. Alice publishes algorithm
+
+For this step, there are some prerequisites needed. If you want to replace the sample algorithm with an algorithm of your choosing, you will need to do some dependency management.
+You can use one of the standard [OCEAN algo_dockers images](https://github.com/oceanprotocol/algo_dockers) or publish a custom docker image.
```
Comment (trentmc) [inline READMEs/c2d-flow.md:196]: "OCEAN" -> "Ocean"

**Pull request merged**